scheduler/rank.go

```diff
@@ -467,36 +469,49 @@ OUTER:
 				}
 			}
 
-			// Check if we need to allocate any reserved cores
-			if task.Resources.Cores > 0 {
-				// set of reservable CPUs for the node
-				nodeCPUSet := cpuset.New(option.Node.NodeResources.Cpu.ReservableCpuCores...)
-				// set of all reserved CPUs on the node
-				allocatedCPUSet := cpuset.New()
-				for _, alloc := range proposed {
-					allocatedCPUSet = allocatedCPUSet.Union(cpuset.New(alloc.ComparableResources().Flattened.Cpu.ReservedCores...))
+			// Handle CPU core reservations
+			if wantedCores := task.Resources.Cores; wantedCores > 0 {
+				// set of cores on this node allowable for use by nomad
+				nodeCores := option.Node.NodeResources.Processors.Topology.UsableCores()
+
+				// set of consumed cores on this node
+				consumedCores := idset.Empty[hw.CoreID]()
+				for _, alloc := range proposed { // proposed is existing + proposal
+					allocCores := alloc.AllocatedResources.Comparable().Flattened.Cpu.ReservedCores
+					idset.InsertSlice(consumedCores, allocCores...)
 				}
 
-				// add any cores that were reserved for other tasks
+				// add cores reserved for other tasks
 				for _, tr := range total.Tasks {
-					allocatedCPUSet = allocatedCPUSet.Union(cpuset.New(tr.Cpu.ReservedCores...))
+					taskCores := tr.Cpu.ReservedCores
+					idset.InsertSlice(consumedCores, taskCores...)
 				}
```

Comment:
The `tr.Cpu.ReservedCores` is an ID set and not a count, right? So what prevents proposed allocs from colliding with existing allocs when they get inserted into the `consumedCores` slice?

(But this would be a problem with the existing code too, so presumably I'm just wrong here.)

Reply:
It's the "that have been placed as part of this same evaluation" that I was missing here. So those have already safely picked cores that weren't in use. And it isn't until below this line that we're proposing an additional set of cores for the alloc we're trying to find a spot for. LGTM :+1: